enumerable/concat.go

```diff
@@ -8,29 +18,49 @@ type enumerableConcat[T any] struct {
 // Concat takes zero to many source `Ènumerable`s and stacks them on top
 // of each other, resulting in one enumerable that will iterate through all
 // the values in all of the given sources.
-func Concat[T any](sources ...Enumerable[T]) Enumerable[T] {
+//
+// New sources may be added after iteration has begun.
+func Concat[T any](sources ...Enumerable[T]) Concatenation[T] {
 	return &enumerableConcat[T]{
 		sources:            sources,
 		currentSourceIndex: 0,
 	}
 }
 
+// Append appends a new source to this concatenation.
+//
+// This may be done after enumeration has begun.
+func (s *enumerableConcat[T]) Append(newSource Enumerable[T]) {
+	s.sources = append(s.sources, newSource)
+}
+
 func (s *enumerableConcat[T]) Next() (bool, error) {
+	startSourceIndex := s.currentSourceIndex
+	hasLooped := false
+
 	for {
```

Comment:
thought: I messed up this logic briefly and there is a strong case for adding `concat` tests within this PR I think as this type is now a bit more complicated.  It is tested by defra however.  Let me know, otherwise I will just see how I feel once approved.

Reply:
It would be nice to have it tested within the repo.

Reply:
I feel like this `Next` logic is a lot more complicated then it needs to be, and a few potential oddities, holding off specifics as you mentioned it was briefly messed up, is the current version the *corrected* version.

To make sure we're on the same page on the intended behavior of `Concatenation` types, can you add a docstring to the interface describing the behavior of the new one. I think its straight forward, but just want to confirm, and make sure others that come across this also know.

Reply:
> I feel like this Next logic is a lot more complicated then it needs to be

The complication involved is to preserve the order of items, so that Next does not bounce in and out of sources.  We want to yield items from source A, then B, then C, then check if A has new items. 

> To make sure we're on the same page on the intended behavior of Concatenation types, can you add a docstring to the interface describing the behavior of the new one.

I'm not sure I understand you here.  There is only one concrete Concatenation type, and one interface, and they are both doucmented.